File: 117.填充每个节点的下一个右侧节点指针-ii.py
#
# @lc app=leetcode.cn id=117 lang=python3
#
# [117] 填充每个节点的下一个右侧节点指针 II
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
from collections import deque


class Solution:
    # 我的
    # def connect(self, root: 'Node') -> 'Node':
    #     if not root:
    #         return None
    #     queue = deque()
    #     queue.append(root)
    #     queue.append(None)
    #     while len(queue) != 0:
    #         node = queue.popleft()
    #         if node:
    #             node.next = queue[0]
    #             if node.left:
    #                 queue.append(node.left)
    #             if node.right:
    #                 queue.append(node.right)
    #         else:
    #             if len(queue) != 0:
    #                 queue.append(None)
    #     return root
    def processChild(self, childNode, prev, leftmost):
        if childNode:
            if prev:
                prev.next = childNode
            else:
                leftmost = childNode
            prev = childNode
        return prev, leftmost

    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root
        leftmost = root
        while leftmost:
            prev, curr = None, leftmost
            leftmost = None
            while curr:
                prev, leftmost = self.processChild(curr.left, prev, leftmost)
                prev, leftmost = self.processChild(curr.right, prev, leftmost)
                curr = curr.next
        return root
# @lc code=end

File: 117.填充每个节点的下一个右侧节点指针-ii.java
/*
 * @lc app=leetcode.cn id=117 lang=java
 *
 * [117] 填充每个节点的下一个右侧节点指针 II
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/
class Solution {
    private Node prev = null;
    private Node leftMost = null;
    public Node connect(Node root) {
        if (root == null)
            return root;
        leftMost = root;
        while (leftMost != null) {
            Node curr = leftMost;
            prev = null;
            leftMost = null;
            while (curr != null) {
                helper(curr.left);
                helper(curr.right);
                curr = curr.next;
            }
        }
        return root;
    }
    private void helper(Node child) {
        if (child != null) {
            if (prev != null)
                prev.next = child;
            else 
                leftMost = child;
            
            prev = child;
        }
    }
}
// @lc code=end

